# Link github: https://github.com/nightfury1103/P4DS-CQ19-FinalProject

# Import

In [3]:
import numpy as np
import pandas as pd

# Thu thập dữ liệu

## Chủ đề dữ liệu

Dữ liệu về sóng thần tính từ năm 1800 đến năm 2021

## Nguồn dữ liệu - Quyền sử dụng

- Dữ liệu sử dụng của notebook được public trên trang web kaggle với đường dẫn sau [link](https://www.kaggle.com/ramjasmaurya/historical-data-of-tsunamis18002021).
- Về vấn đề quyền sử:
    Tác giả đề cập đến vấn đề quyền sử dụng dựa theo [CC0 1.0 Universal (CC0 1.0) Public Domain Dedication](https://creativecommons.org/publicdomain/zero/1.0/)

## Cách thu thập dữ liệu

Để đơn giản trong đồ án này dữ liệu được tải từ trang với đường dẫn ở trên. Sau đó file được lưu vào folder `Data` nhằm mục đích sử dụng. 

## Mô tả dữ liệu

DATASET gives brief description of Tsunamis that were happened between the years 1800 to 2021.

Dataset has 26 columns all including from date metrics to measure metrics with its damages.

Some info:

Tsunami Event Database contains information on the source of the tsunami. It is related to the Tsunami Runup database which contains information on locations where tsunami effects occurred.

Tsunami Runup Database contains information on locations where tsunami effects occurred. It is related to the Tsunami Source Event database which contains information on the source of the tsunami.

Tsunami deposits are the physical evidence left behind when a tsunami impacts a shoreline or affects submarine sediments. These deposits can be preserved in the geologic record and studied long after the tsunami waves dissipate.

tsunami marigrams from analog (microfiche) to high-resolution digital images. Microfiche were created in the 1970s from deteriorating paper records. The collection includes records from 1854 to 1981.

# Khám phá dữ liệu

## Đọc dữ liệu

In [22]:
tsunami_df = pd.read_csv('Data/tsunami.csv')
tsunami_df.head()

Year   Mo    Dy    Hr    Mn  Sec  Tsunami Event Validity  \
0     NaN  NaN   NaN   NaN   NaN  NaN                     NaN   
1  1800.0  6.0   2.0  20.0  17.0  NaN                     2.0   
2  1802.0  1.0   4.0   7.0   NaN  NaN                     2.0   
3  1802.0  3.0  19.0   NaN   NaN  NaN                     2.0   
4  1802.0  8.0  10.0   NaN   NaN  NaN                     1.0   

   Tsunami Cause Code  Earthquake Magnitude  Deposits  ...  \
0                 NaN                   NaN       NaN  ...   
1                 1.0                   NaN       0.0  ...   
2                 1.0                   NaN       0.0  ...   
3                 1.0                   NaN       0.0  ...   
4                 0.0                   NaN       0.0  ...   

  Tsunami Magnitude (Iida) Tsunami Intensity  Total Deaths  Total Missing  \
0                      NaN               NaN           NaN            NaN   
1                      NaN               NaN           NaN            NaN   
2                      NaN               3.0           NaN            NaN   
3                      NaN               NaN           NaN            NaN   
4                      NaN               NaN           NaN            NaN   

   Total Missing Description  Total Injuries  Total Damage ($Mil)  \
0                        NaN             NaN                  NaN   
1                        NaN             NaN                  NaN   
2                        NaN             NaN                  NaN   
3                        NaN             NaN                  NaN   
4                        NaN             NaN                  NaN   

   Total Damage Description  Total Houses Destroyed  Total Houses Damaged  
0                       NaN                     NaN                   NaN  
1                       NaN                     NaN                   NaN  
2                       NaN                     NaN                   NaN  
3                       NaN                     NaN                   NaN  
4                       NaN                     NaN                   NaN  

[5 rows x 26 columns]

## Dữ liệu có bao nhiêu dòng và cột ? 

Dựa vào dữ liệu đưa vào ta có thể thấy dữ liệu bị trống ở dòng đầu tiên. Do đó, ta sẽ thực hiệp xóa dòng này trước khi khám phá dữ liệu

In [23]:
tsunami_df = tsunami_df.drop(0, axis=0)

In [24]:
tsunami_df.shape

(2161, 26)

Ta thấy dữ liệu có 2161 dòng và 26 cột tương ứng

## Các dòng dữ liệu

### Mỗi dòng có ý nghĩa gì? Có vấn đề các dòng có ý nghĩa khác nhau không?

In [33]:
tsunami_df.sample(10)

Year    Mo    Dy    Hr    Mn   Sec  Tsunami Event Validity  \
467   1869.0   8.0  11.0   NaN   NaN   NaN                     1.0   
754   1898.0  12.0   3.0   5.0  50.0   NaN                     2.0   
214   1847.0   7.0   9.0   NaN   NaN   NaN                     4.0   
1484  1965.0   2.0   4.0   5.0   1.0  21.6                     4.0   
1282  1945.0   2.0  10.0   4.0  57.0   0.0                     4.0   
1106  1928.0   5.0  18.0   NaN   NaN   NaN                     2.0   
2126  2018.0  12.0  22.0  13.0  55.0   NaN                     4.0   
274   1854.0   1.0   4.0   NaN   NaN   NaN                     2.0   
549   1877.0   9.0   2.0   NaN   NaN   NaN                     1.0   
1963  2009.0   5.0  28.0   8.0  24.0  45.0                     4.0   

      Tsunami Cause Code  Earthquake Magnitude  Deposits  ...  \
467                  0.0                   NaN       0.0  ...   
754                  1.0                   NaN       0.0  ...   
214                  8.0                   NaN       0.0  ...   
1484                 1.0                   8.7       0.0  ...   
1282                 1.0                   7.3       0.0  ...   
1106                 1.0                   NaN       0.0  ...   
2126                 7.0                   NaN       2.0  ...   
274                  1.0                   6.0       0.0  ...   
549                  0.0                   NaN       0.0  ...   
1963                 1.0                   7.3       0.0  ...   

     Tsunami Magnitude (Iida) Tsunami Intensity  Total Deaths  Total Missing  \
467                       NaN              -1.0           NaN            NaN   
754                       NaN               3.0           NaN            NaN   
214                       NaN               NaN          10.0          100.0   
1484                      3.3               2.5           NaN            NaN   
1282                     -1.3              -1.5           2.0            NaN   
1106                      NaN               0.5           NaN            NaN   
2126                      NaN               NaN         437.0           10.0   
274                       NaN               0.0           NaN            NaN   
549                       NaN               0.5           NaN            NaN   
1963                      NaN               NaN           7.0            NaN   

      Total Missing Description  Total Injuries  Total Damage ($Mil)  \
467                         NaN             NaN                  NaN   
754                         NaN             NaN                  NaN   
214                         2.0             NaN                  NaN   
1484                        NaN             NaN                 0.01   
1282                        NaN             NaN                  NaN   
1106                        NaN             NaN                  NaN   
2126                        1.0         31943.0               250.00   
274                         NaN             NaN                  NaN   
549                         NaN             NaN                  NaN   
1963                        NaN            40.0                  NaN   

      Total Damage Description  Total Houses Destroyed  Total Houses Damaged  
467                        NaN                     NaN                   NaN  
754                        1.0                     NaN                   NaN  
214                        NaN                     NaN                   NaN  
1484                       1.0                     NaN                   NaN  
1282                       NaN                     2.0                   NaN  
1106                       NaN                     NaN                   NaN  
2126                       4.0                     NaN                2844.0  
274                        NaN                     NaN                   NaN  
549                        NaN                     NaN                   NaN  
1963                       2.0                   135.0   

Mỗi dòng là từng số liệu thống kê sóng thần theo từng năm. Có vẻ như không có vấn đề gì giữa mỗi dòng

### Dữ liệu các dòng có bị lặp hay không?

In [40]:
tsunami_df.describe().T

count          mean            std       min  \
Year                       2161.0   1925.722814      58.996433  1800.000   
Mo                         2112.0      6.589015       3.477291     1.000   
Dy                         2054.0     15.531646       8.602083     1.000   
Hr                         1338.0     11.294469       6.983150     0.000   
Mn                         1289.0     28.356090      17.235525     0.000   
Sec                         941.0     26.556642      18.091679     0.000   
Tsunami Event Validity     2161.0      2.490051       1.526318    -1.000   
Tsunami Cause Code         2158.0      1.946247       2.419570     0.000   
Earthquake Magnitude       1293.0      7.027920       0.777719     3.700   
Deposits                   2161.0      0.241092       3.322022     0.000   
Latitude                   1892.0     13.685687      28.811748   -63.000   
Longitude                  1892.0     37.707962     110.683437  -179.971   
Maximum Water Height (m)   1109.0      5.706186      26.737880     0.010   
Number of Runups           2161.0     13.172605     149.375885     0.000   
Tsunami Magnitude (Iida)    576.0      0.448576       2.212544    -4.640   
Tsunami Intensity           897.0      1.032397       1.792976    -4.140   
Total Deaths                492.0   3271.034553   20341.118896     1.000   
Total Missing                11.0   4144.545455   13047.746567     1.000   
Total Missing Description    13.0      2.076923       1.115164     1.000   
Total Injuries              281.0   5247.145907   30604.662566     1.000   
Total Damage ($Mil)         143.0   4197.016455   21700.862491     0.003   
Total Damage Description    907.0      2.093716       1.070905     1.000   
Total Houses Destroyed      197.0  41260.355330  385219.385556     1.000   
Total Houses Damaged         77.0  28113.350649   74978.553547     2.000   

                                 25%        50%          75%          max  
Year                       1877.0000  1927.0000   1977.00000     2021.000  
Mo                            4.0000     7.0000     10.00000       12.000  
Dy                            8.0000    16.0000     23.00000       31.000  
Hr                            5.0000    11.0000     18.00000       23.000  
Mn                           14.0000    29.0000     43.00000       59.000  
Sec                          10.8000    25.7000     42.00000       59.900  
Tsunami Event Validity        1.0000     3.0000      4.00000        4.000  
Tsunami Cause Code            1.0000     1.0000      1.00000       11.000  
Earthquake Magnitude          6.5000     7.0000      7.60000        9.500  
Deposits                      0.0000     0.0000      0.00000      144.000  
Latitude                     -8.3050    16.0000     38.39150       71.813  
Longitude                   -71.5000    50.6745    140.33475      179.984  
Maximum Water Height (m)      0.2000     1.0000      3.50000      524.600  
Number of Runups              1.0000     1.0000      3.00000     6288.000  
Tsunami Magnitude (Iida)     -1.0000     0.7000      2.00000        8.400  
Tsunami Intensity             0.0000     1.0000      2.00000        9.000  
Total Deaths                  4.0000    24.5000    262.75000   316000.000  
Total Missing                20.0000   100.0000    498.00000    43476.000  
Total Missing Description     1.0000     2.0000      3.00000        4.000  
Total Injuries               16.0000   100.0000    500.00000   374171.000  
Total Damage ($Mil)           2.4125    39.0000    520.00000   220136.600  
Total Damage Description      1.0000     2.0000      3.00000        4.000  
Total Houses Destroyed       50.0000   588.0000   5043.00000  5360000.000  
Total Houses Damaged        127.0000  1000.0000  10044.00000   500000.000

Dựa vào số năm là 2160 thì có vẻ như dữ liệu không có dòng nào bị lặp.

## Các cột dữ liệu

### Ý nghĩa các cột

Như đã nói ở trên dữ liệu có tất cả 26 cột ta sẽ tìm hiểu ý nghĩa của các cột ngay sau đây

|Column|Description|
|----------- |----------- |
|Year|Thời điểm xảy ra sóng thần đơn vị năm.|
|Mo|Thời điểm xảy ra sóng thần đơn vị tháng.|
|Dy|Thời điểm xảy ra sóng thần đơn vị ngày.|
|Hr|Thời điểm xảy ra sóng thần đơn vị giờ.|
|Mn|Thời điểm xảy ra sóng thần đơn vị phút.|
|Sec|Thời điểm xảy ra sóng thần đơn vị giây.|
|Tsunami Event Validity|Độ chuẩn xác của sóng thần|
|Tsunami Cause Code|Mã nguyên nhân xảy ra sóng thần.|
|Earthquake Magnitude|Tầm ảnh hưởng của động đất.|
|Deposits|Trầm tích sóng thần(trầm tích lắng đọng do kết quả của sóng thần).|
|Country|Đất nước xảy ra sóng thần.|
|Location Name|Địa điểm xảy ra sóng thần.|
|Latitude|Vĩ độ xảy ra sóng thần.|
|Longitude|Kinh độ xảy ra sóng thần.|
|Maximum Water Height (m)|Độ cao đạt đỉnh của sóng thần đơn vị mét.|
|Number of Runups|Số lần chạy(tăng cao) của sóng thần|
|Tsunami Magnitude (Iida)|Kích cỡ của sóng thần|
|Tsunami Intensity|Cường độ của sóng thần|
|Total Deaths|Tổng số người chết|
|Total Missing|Tổng số người mất tích|
|Total Missing Description|Tổng số người mất tích mô tả|
|Total Injuries|Tổng số người bị thương|
|Total Damage ($Mil)|Tổng số thiệt hại|
|Total Damage Description|Tổng số thiệt hại mô tả|
|Total Houses Destroyed|Tổng số ngôi nhà bị phá hủy|
|Total Houses Damaged|Tổng số ngôi nhà bị hư hại|

### Mỗi cột hiện đang có kiểu dữ liệu gì? Có cột nào có kiểu dữ liệu chưa phù hợp để có thể xử lý tiếp không?

In [38]:
tsunami_df.dtypes

Year                         float64
Mo                           float64
Dy                           float64
Hr                           float64
Mn                           float64
Sec                          float64
Tsunami Event Validity       float64
Tsunami Cause Code           float64
Earthquake Magnitude         float64
Deposits                     float64
Country                       object
Location Name                 object
Latitude                     float64
Longitude                    float64
Maximum Water Height (m)     float64
Number of Runups             float64
Tsunami Magnitude (Iida)     float64
Tsunami Intensity            float64
Total Deaths                 float64
Total Missing                float64
Total Missing Description    float64
Total Injuries               float64
Total Damage ($Mil)          float64
Total Damage Description     float64
Total Houses Destroyed       float64
Total Houses Damaged         float64
dtype: object

Có vẻ như mỗi cột điều có kiểu dữ liệu phù hợp.

### Với mỗi cột có kiểu dữ liệu dạng số (numerical), các giá trị được phân bố như thế nào

In [42]:
tsunami_df.describe().T

count          mean            std       min  \
Year                       2161.0   1925.722814      58.996433  1800.000   
Mo                         2112.0      6.589015       3.477291     1.000   
Dy                         2054.0     15.531646       8.602083     1.000   
Hr                         1338.0     11.294469       6.983150     0.000   
Mn                         1289.0     28.356090      17.235525     0.000   
Sec                         941.0     26.556642      18.091679     0.000   
Tsunami Event Validity     2161.0      2.490051       1.526318    -1.000   
Tsunami Cause Code         2158.0      1.946247       2.419570     0.000   
Earthquake Magnitude       1293.0      7.027920       0.777719     3.700   
Deposits                   2161.0      0.241092       3.322022     0.000   
Latitude                   1892.0     13.685687      28.811748   -63.000   
Longitude                  1892.0     37.707962     110.683437  -179.971   
Maximum Water Height (m)   1109.0      5.706186      26.737880     0.010   
Number of Runups           2161.0     13.172605     149.375885     0.000   
Tsunami Magnitude (Iida)    576.0      0.448576       2.212544    -4.640   
Tsunami Intensity           897.0      1.032397       1.792976    -4.140   
Total Deaths                492.0   3271.034553   20341.118896     1.000   
Total Missing                11.0   4144.545455   13047.746567     1.000   
Total Missing Description    13.0      2.076923       1.115164     1.000   
Total Injuries              281.0   5247.145907   30604.662566     1.000   
Total Damage ($Mil)         143.0   4197.016455   21700.862491     0.003   
Total Damage Description    907.0      2.093716       1.070905     1.000   
Total Houses Destroyed      197.0  41260.355330  385219.385556     1.000   
Total Houses Damaged         77.0  28113.350649   74978.553547     2.000   

                                 25%        50%          75%          max  
Year                       1877.0000  1927.0000   1977.00000     2021.000  
Mo                            4.0000     7.0000     10.00000       12.000  
Dy                            8.0000    16.0000     23.00000       31.000  
Hr                            5.0000    11.0000     18.00000       23.000  
Mn                           14.0000    29.0000     43.00000       59.000  
Sec                          10.8000    25.7000     42.00000       59.900  
Tsunami Event Validity        1.0000     3.0000      4.00000        4.000  
Tsunami Cause Code            1.0000     1.0000      1.00000       11.000  
Earthquake Magnitude          6.5000     7.0000      7.60000        9.500  
Deposits                      0.0000     0.0000      0.00000      144.000  
Latitude                     -8.3050    16.0000     38.39150       71.813  
Longitude                   -71.5000    50.6745    140.33475      179.984  
Maximum Water Height (m)      0.2000     1.0000      3.50000      524.600  
Number of Runups              1.0000     1.0000      3.00000     6288.000  
Tsunami Magnitude (Iida)     -1.0000     0.7000      2.00000        8.400  
Tsunami Intensity             0.0000     1.0000      2.00000        9.000  
Total Deaths                  4.0000    24.5000    262.75000   316000.000  
Total Missing                20.0000   100.0000    498.00000    43476.000  
Total Missing Description     1.0000     2.0000      3.00000        4.000  
Total Injuries               16.0000   100.0000    500.00000   374171.000  
Total Damage ($Mil)           2.4125    39.0000    520.00000   220136.600  
Total Damage Description      1.0000     2.0000      3.00000        4.000  
Total Houses Destroyed       50.0000   588.0000   5043.00000  5360000.000  
Total Houses Damaged        127.0000  1000.0000  10044.00000   500000.000

Có vẻ như các dữ liệu dạng số ở phần nửa trên tương đối đầy đủ so với nửa dưới.

#### Số-lượng/tỉ-lệ các giá trị thiếu?

In [44]:
nume_col_info_df = pd.DataFrame()
for col in tsunami_df.keys():
    if tsunami_df[col].dtype in [np.float32, np.float64]:
        missing_percentage = round(tsunami_df[col].isnull().sum() * 100 / len(tsunami_df[col]), 1)
        min_value = tsunami_df[col].min()
        lower_quartile = tsunami_df[col].quantile(0.25)
        median = tsunami_df[col].median()
        upper = tsunami_df[col].quantile(0.75)
        max_value = tsunami_df[col].max()
        row_line = pd.Series([missing_percentage, min_value, lower_quartile, median, upper, max_value], 
                             index = ['missing_percentage', 'min', 'lower_quartile', 'median', 'upper_quartile', 'max'])
        nume_col_info_df[col] = row_line
nume_col_info_df.T

missing_percentage       min  lower_quartile  \
Year                                      0.0  1800.000       1877.0000   
Mo                                        2.3     1.000          4.0000   
Dy                                        5.0     1.000          8.0000   
Hr                                       38.1     0.000          5.0000   
Mn                                       40.4     0.000         14.0000   
Sec                                      56.5     0.000         10.8000   
Tsunami Event Validity                    0.0    -1.000          1.0000   
Tsunami Cause Code                        0.1     0.000          1.0000   
Earthquake Magnitude                     40.2     3.700          6.5000   
Deposits                                  0.0     0.000          0.0000   
Latitude                                 12.4   -63.000         -8.3050   
Longitude                                12.4  -179.971        -71.5000   
Maximum Water Height (m)                 48.7     0.010          0.2000   
Number of Runups                          0.0     0.000          1.0000   
Tsunami Magnitude (Iida)                 73.3    -4.640         -1.0000   
Tsunami Intensity                        58.5    -4.140          0.0000   
Total Deaths                             77.2     1.000          4.0000   
Total Missing                            99.5     1.000         20.0000   
Total Missing Description                99.4     1.000          1.0000   
Total Injuries                           87.0     1.000         16.0000   
Total Damage ($Mil)                      93.4     0.003          2.4125   
Total Damage Description                 58.0     1.000          1.0000   
Total Houses Destroyed                   90.9     1.000         50.0000   
Total Houses Damaged                     96.4     2.000        127.0000   

                              median  upper_quartile          max  
Year                       1927.0000      1977.00000     2021.000  
Mo                            7.0000        10.00000       12.000  
Dy                           16.0000        23.00000       31.000  
Hr                           11.0000        18.00000       23.000  
Mn                           29.0000        43.00000       59.000  
Sec                          25.7000        42.00000       59.900  
Tsunami Event Validity        3.0000         4.00000        4.000  
Tsunami Cause Code            1.0000         1.00000       11.000  
Earthquake Magnitude          7.0000         7.60000        9.500  
Deposits                      0.0000         0.00000      144.000  
Latitude                     16.0000        38.39150       71.813  
Longitude                    50.6745       140.33475      179.984  
Maximum Water Height (m)      1.0000         3.50000      524.600  
Number of Runups              1.0000         3.00000     6288.000  
Tsunami Magnitude (Iida)      0.7000         2.00000        8.400  
Tsunami Intensity             1.0000         2.00000        9.000  
Total Deaths                 24.5000       262.75000   316000.000  
Total Missing               100.0000       498.00000    43476.000  
Total Missing Description     2.0000         3.00000        4.000  
Total Injuries              100.0000       500.00000   374171.000  
Total Damage ($Mil)          39.0000       520.00000   220136.600  
Total Damage Description      2.0000         3.00000        4.000  
Total Houses Destroyed      588.0000      5043.00000  5360000.000  
Total Houses Damaged       1000.0000     10044.00000   500000.000

#### Min? max? → Có gì bất thường không?

Dựa vào dataframe ở trên thì có vẻ như min max không có gì bất thường ở dữ liệu này.

### Với mỗi cột có kiểu dữ liệu dạng phân loại (categorical), các giá trị được phân bố như thế nào?

#### Số-lượng/tỉ-lệ các giá trị thiếu?

In [64]:
cate_col_info_df = pd.DataFrame()
for col in tsunami_df.keys():
    if tsunami_df[col].dtype not in [np.float32, np.float64]:
        missing_percentage = round(tsunami_df[col].isnull().sum() * 100 / len(tsunami_df[col]), 1)
#         if col in multichoice_cols:
#             x = tsunami_df[col].str.split(';')
#             num_values = len(x.explode().value_counts())
#             value_percentages = round(x.explode().value_counts() * 100/x.explode().value_counts().sum(), 1)
#         else:
        num_values = len(tsunami_df[col].value_counts())
        value_percentages = round(tsunami_df[col].value_counts() * 100/tsunami_df[col].notnull().sum(), 1)
        cate_col_info_df[col] = pd.Series([missing_percentage, num_values, value_percentages], 
                                          index=['missing_percentage', 'num_values', 'value_percentages'])
cate_col_info_df

Country  \
missing_percentage                                                0.0   
num_values                                                        104   
value_percentages   USA          12.0
INDONESIA    10.6
JAPAN     ...   

                                                        Location Name  
missing_percentage                                                0.1  
num_values                                                        797  
value_percentages   NORTHERN CHILE                3.3
CENTRAL CHIL...

#### Số lượng các giá trị khác nhau? Show một vài giá trị → Có gì bất thường không?

In [87]:
country_df = pd.DataFrame(cate_col_info_df['Country']['value_percentages'])
country_df.head(5)

Country
USA           12.0
INDONESIA     10.6
JAPAN         10.4
CHILE          7.0
GREECE         5.2

In [88]:
location_df = pd.DataFrame(cate_col_info_df['Location Name']['value_percentages'])
location_df.head(5)

Location Name
NORTHERN CHILE              3.3
CENTRAL CHILE               2.5
HAWAII                      1.9
SANRIKU                     1.7
S. KURIL ISLANDS            1.6

Có vẻ dữ liệu đều ổn không có gì bất thường xảy ra ở đây

# Đưa ra các câu hỏi có ý nghĩa cần trả lời

## Câu hỏi - Lợi ích

# Tiền xử lý + phân tích dữ liệu để trả lời cho từng câu hỏi

# Nhìn lại quá trình làm đồ án

## Những khó khăn

## Học được gì hữu ích

## Nếu có thêm thời gian sẽ làm gì

# Tài liệu tham khảo